logstash-core/lib/logstash/settings.rb

```diff
@@ -148,6 +148,16 @@ def flatten_hash(h,f="",g={})
       end
       g
     end
+
+    # Substitutes environment variables into setting values of string type.
+    def substitute_env(settings)
+      settings.each do |_, setting|
+        if setting.is_a? String
```

Comment:
@original-brownbear I don't think they will be, because settings will return `Array` instead of `String`, I think we need to make the algorithm visit the child node.